mava/networks/utils/mat/decode.py

```diff
@@ -0,0 +1,171 @@
+
+    masked_logits = jnp.where(
+        legal_actions,
+        logit,
+        jnp.finfo(jnp.float32).min,
+    )
+
+    distribution = IdentityTransformation(distribution=tfd.Categorical(logits=masked_logits))
+    action_log_prob = distribution.log_prob(action)
+    entropy = distribution.entropy(seed=key)
+
+    return action_log_prob, entropy  # (B, N), (B, N)
+
+
+def continuous_parallel_act(
+    decoder: nn.Module,
+    obs_rep: chex.Array,  # (B, N, E)
+    action: chex.Array,  # (B, N, A)
+    action_dim: int,  # (, )
+    legal_actions: chex.Array,  # (B, N, A)
+    key: chex.PRNGKey,
+) -> Tuple[chex.Array, chex.Array]:
+    # We don't need legal_actions for continuous actions but keep it to keep the APIs consistent.
+    del legal_actions
+    batch_size, n_agent, _ = obs_rep.shape
+    shifted_action = jnp.zeros((batch_size, n_agent, action_dim))
+
+    shifted_action = shifted_action.at[:, 1:, :].set(action[:, :-1, :])
+
+    act_mean = decoder(shifted_action, obs_rep)  # (B, N, A)
+    action_std = jax.nn.softplus(decoder.log_std)
+
+    distribution = tfd.Normal(loc=act_mean, scale=action_std)
+    distribution = tfd.Independent(
+        TanhTransformedDistribution(distribution),
+        reinterpreted_batch_ndims=1,
+    )
+    action_log_prob = distribution.log_prob(action)
+    entropy = distribution.entropy(seed=key)
+
+    return action_log_prob, entropy  # (B, N), (B, N)
+
+
+def discrete_autoregressive_act(
+    decoder: nn.Module,
+    obs_rep: chex.Array,  # (B, N, E)
+    action_dim: int,  # (, )
+    legal_actions: chex.Array,  # (B, N, A)
+    key: chex.PRNGKey,
+) -> Tuple[chex.Array, chex.Array]:
+    batch_size, n_agent, _ = obs_rep.shape
+    shifted_action = jnp.zeros((batch_size, n_agent, action_dim + 1))
+    shifted_action = shifted_action.at[:, 0, 0].set(1)
+    output_action = jnp.zeros((batch_size, n_agent))
+    output_action_log = jnp.zeros_like(output_action)
+
+    for i in range(n_agent):
+        logit = decoder(shifted_action, obs_rep)[:, i, :]  # (B, A)
+        masked_logits = jnp.where(
+            legal_actions[:, i, :],
+            logit,
+            jnp.finfo(jnp.float32).min,
+        )
+        key, sample_key = jax.random.split(key)
+
+        distribution = IdentityTransformation(distribution=tfd.Categorical(logits=masked_logits))
+        action = distribution.sample(seed=sample_key)  # (B, )
+        action_log = distribution.log_prob(action)  # (B, )
+
+        output_action = output_action.at[:, i].set(action)
+        output_action_log = output_action_log.at[:, i].set(action_log)
+
+        update_shifted_action = i + 1 < n_agent
+        shifted_action = jax.lax.cond(
+            update_shifted_action,
+            lambda action=action, i=i, shifted_action=shifted_action: shifted_action.at[
+                :, i + 1, 1:
+            ].set(jax.nn.one_hot(action, action_dim)),
+            lambda shifted_action=shifted_action: shifted_action,
+        )
```

Comment:
I think we can just do an `at[].set()` [with mode = drop](https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html)

```suggestion
        # Adds all expect the last action to shifted_actions, as it is out of range
        shifted_action.at[:, i + 1, 1:].set(jax.nn.one_hot(action, action_dim), mode="drop")
```

Reply:
Done.